Add OLC (pluscode) library and its element to OSD

giacomo892 opened the pull request:
This PR adds the AP_OLC library that allows to generate OLC (open location codes) from a give latitude and longitude pair and  a new OSD element, PLUSCODE, that allow to display the +code on the OSD screen using less space than the current lat/lon pair.
More info on: https://github.com/google/open-location-code

Thanks @shellixyz for the help and @fiam for the original code in INAV.

Gone4Dirt commented:
Interesting.  Forgive my ignorance but what is the use case here?  I am curious, other than reduced space on the OSD what does using OLC give you over using lat long.

giacomo892 commented:
> Interesting. Forgive my ignorance but what is the use case here? I am curious, other than reduced space on the OSD what does using OLC give you over using lat long.

It uses much less space on the screen

giacomo892 commented:
![image](https://user-images.githubusercontent.com/20860004/91726443-88c30b80-eba0-11ea-9bd0-77cca5b2fdc8.png)
Tested on SITL, you can see in the bottom part of the OSD which are the benefits of this PR

peterbarker commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,226 @@
+// to avoid double floating point math and use integer math as much as possible.
+
+static constexpr char SEPARATOR_CHAR = '+';
+static constexpr int SEPARATOR_POS = 8;
```
Please stick to stdint types as much as possible - e.g. `uint8_t` here.

peterbarker commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,226 @@
+olc_coord_t AP_OLC::initial_resolution;
+const char AP_OLC::alphabet[] = "23456789CFGHJMPQRVWX";
+
+bool AP_OLC::inited = false;
```
No need for this to be a static.

peterbarker commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,226 @@
+    inited = true;
+
+    // Work out the encoding base exponent necessary to represent 360 degrees.
+    olc_coord_t initial_exponent = floorf(logf(2 * (LON_MAX / OLC_DEG_MULTIPLIER)) / logf(ENCODING_BASE));
```
These looks like a constant to me.

They ought to be calculated ahead of time?

peterbarker commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,50 @@
+// olc_encodes the given coordinates in lat and lon (deg * OLC_DEG_MULTIPLIER)
+// as an OLC code of the given length. It returns the number of characters
+// written to buf.
+static int olc_encode(olc_coord_t lat, olc_coord_t lon, size_t length, char *buf, size_t bufsize);
```
Indenting needs fixing.  I suggest running our astyle tool over this.

peterbarker replied:
External interfaces should probably be using type of what we expect.  We'll just be throwing `int32_t`s in...

peterbarker commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,50 @@
+ */
+
+#pragma once
+#include <AP_Common/AP_Common.h>
```
What are we using out of this?

peterbarker commented on libraries/AP_OSD/AP_OSD.h:
```diff
@@ -127,7 +127,8 @@ class AP_OSD_Screen
     AP_OSD_Setting clk{false, 0, 0};
-    
+    AP_OSD_Setting pluscode{false, 0, 0};
+
```
New whitespace?

giacomo892 replied:
Actually there is none added

peterbarker commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -706,6 +706,22 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
     
+    // @Param: PLUSCODE_EN
+    // @DisplayName: PLUSCODE_EN
+    // @Description: Displays total flight time
```
This parameter description needs some help...

peterbarker reviewed (changes requested):
I haven't delved into the mechanics of this yet.

I'm not sure the "why" question has been adequately answered.  Is the proposal that users can stop displaying lat/lng and start displaying the pluscode instead?  It looks like it fulfils the same role - does the pluscode implementation here give as much precision as the replaced lat/lon?

Gone4Dirt commented:
Just to add to my original question.  I don't understand what the use case for displaying position on OSD is.  whether is it lat/lng or OLC.  I just don't see how it is useful to see position as a number/code on an OSD.  Happy to be corrected, I just feel that it needs to a have reasonable usecase, and its not clear to me what that is.  So could you explain the use case please?

giacomo892 commented:
Hi @peterbarker thanks for your review. I should have addressed your concerns. Let me know what do you think.

Regarding the usefulness of this PR (so I can reply to @Gone4Dirt too) I think it is nice to have. Won't add any precision but it allows to display the position of the machine on the OSD using just less space (half of it?) than the LAT/LON combo. 
I think having the position displayed on the OSD is always useful as it gives another way to find the machine in case of an unfortunate crash.

peterbarker commented:
Marked this for DevCallEU, @andyp1per

andyp1per commented:
Seems like a lot of code for a feature I don't think I would use. Is this coding quite common? I agree that printing lat/lng is mostly useless - unless you crash a long way away.

tridge commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,225 @@
+
+olc_coord_t AP_OLC::grid_size;
+olc_coord_t AP_OLC::initial_resolution;
+char alphabet[] = "23456789CFGHJMPQRVWX";
```
const and not global

tridge commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include <stdint.h>
+
+typedef int32_t olc_coord_t;
```
just use int32_t, like we use elsewhere

tridge commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,49 @@
+
+    static void init_constants(void);
+    static float compute_precision_for_length(int length);
+    static olc_coord_t adjust_latitude(olc_coord_t lat, size_t code_len);
```
use int32_t

tridge reviewed (changes requested):
please find flash cost, and look at checking for HAL_MINIMIZE_FEATURES, cleanup namespace usage

giacomo892 commented:
@tridge I should have addressed your review.
The flash usage is 816bytes on F405-STD comparing this branch with current master

tridge commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,43 @@
+#pragma once
+#include <stdint.h>
+
+typedef int32_t int32_t;
```
odd typedef ....

andyp1per commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,214 @@
+
+namespace {
+
+    constexpr uint8_t SEPARATOR_CHAR = '+';
```
I think these should be in the class

andyp1per commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,214 @@
+
+    const char alphabet[] = "23456789CFGHJMPQRVWX";
+
+    int32_t initial_exponent = floorf(logf(2 * (LON_MAX / OLC_DEG_MULTIPLIER)) / logf(ENCODING_BASE));
```
these should be const

tridge commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,43 @@
+
+typedef int32_t int32_t;
+
+
```
add in this:
```
#ifndef HAL_PLUSCODE_ENABLE
#define HAL_PLUSCODE_ENABLE !HAL_MINIMIZE_FEATURES && BOARD_FLASH_SIZE > 1024
#endif
```
then use in AP_OSD and the class definition

tridge commented on libraries/AP_OLC/AP_OLC.h:
```diff
@@ -0,0 +1,43 @@
+typedef int32_t int32_t;
+
+
+static constexpr int32_t OLC_DEG_MULTIPLIER = 10000000; // 1e7
```
move that const to inside the AP_OLC class

tridge commented on libraries/AP_OSD/AP_OSD_Screen.cpp:
```diff
@@ -1690,6 +1705,19 @@ void AP_OSD_Screen::draw_clk(uint8_t x, uint8_t y)
 
+void AP_OSD_Screen::draw_pluscode(uint8_t x, uint8_t y)
+{
+    AP_GPS & gps = AP::gps();
```
use
#if HAL_PLUSCODE_ENABLE for this

andyp1per commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,214 @@
+
+int32_t AP_OLC::adjust_latitude(int32_t lat, size_t code_len)
+{
+    if (lat < -LAT_MAX) {
```
use constrain_int32_t here

peterbarker commented on libraries/AP_OLC/AP_OLC.cpp:
```diff
@@ -0,0 +1,190 @@
+// Compute the latitude precision value for a given code length.  Lengths <= 10
+// have the same precision for latitude and longitude, but lengths > 10 have
+// different precisions due to the grid method having fewer columns than rows.
+float AP_OLC::compute_precision_for_length(int length)
```
This should be unsigned; there are other cases of non-stdint type in here, too.

peterbarker commented:
Merged, thanks!

amilcarlucas commented:
How do I set the HAL_PLUSCODE_ENABLE to false? is it a waf command line option?

giacomo892 commented:
It's off by default on boards with 1MB flash. To enable it just set HAL_PLUSCODE_ENABLE 1 in the hwdef
